File: solutions/typescript/state-of-tic-tac-toe/1/state-of-tic-tac-toe.ts
type Player = 'X' | 'O';
export type BoardState = 'ongoing' | 'draw' | 'win';

export const gamestate = (board: string[]): BoardState => {
  // 1. Basic validation of board shape
  if (!board || board.length !== 3 || board.some(row => row.length !== 3)) {
    throw new Error('Invalid board dimensions');
  }

  // Count X and O placements
  let xCount = 0;
  let oCount = 0;
  for (const row of board) {
    for (const cell of row) {
      if (cell === 'X') xCount++;
      else if (cell === 'O') oCount++;
    }
  }

  // 2. Validate Turn Order with precise error messages
  if (xCount > oCount + 1) {
    throw new Error('Wrong turn order: X went twice');
  }
  if (xCount < oCount) {
    throw new Error('Wrong turn order: O started');
  }

  // 3. Helper to check if a specific player has 3-in-a-row
  const checkWin = (player: Player): boolean => {
    // Rows
    if (board.some(row => row.split('').every(c => c === player))) return true;
    
    // Columns
    for (let col = 0; col < 3; col++) {
      if (board[0][col] === player && board[1][col] === player && board[2][col] === player) {
        return true;
      }
    }
    
    // Diagonals
    if (board[0][0] === player && board[1][1] === player && board[2][2] === player) return true;
    if (board[0][2] === player && board[1][1] === player && board[2][0] === player) return true;

    return false;
  };

  const xWon = checkWin('X');
  const oWon = checkWin('O');

  // 4. Validate Continued Play After Win
  // Both winning or a player winning with incorrect turn counts mean game continued past a win
  if ((xWon && oWon) || (xWon && xCount !== oCount + 1) || (oWon && xCount !== oCount)) {
    throw new Error('Impossible board: game should have ended after the game was won');
  }

  // 5. Determine and Return State
  if (xWon || oWon) {
    return 'win';
  }

  // If no one won and the board is full
  if (xCount + oCount === 9) {
    return 'draw';
  }

  return 'ongoing';
};

File: solutions/python/state-of-tic-tac-toe/1/state_of_tic_tac_toe.py
"""Determine the state of a tic-tac-toe game."""


def gamestate(board):
    """Return the state of the game or raise ValueError for invalid boards."""
    x_count = sum(row.count("X") for row in board)
    o_count = sum(row.count("O") for row in board)

    if o_count > x_count:
        raise ValueError("Wrong turn order: O started")

    if x_count - o_count > 1:
        raise ValueError("Wrong turn order: X went twice")

    winning_lines = []

    for row_index in range(3):
        winning_lines.append([board[row_index][column_index] for column_index in range(3)])

    for column_index in range(3):
        winning_lines.append([board[row_index][column_index] for row_index in range(3)])

    winning_lines.append([board[index][index] for index in range(3)])
    winning_lines.append([board[index][2 - index] for index in range(3)])

    x_wins = any(line == ["X", "X", "X"] for line in winning_lines)
    o_wins = any(line == ["O", "O", "O"] for line in winning_lines)

    if x_wins and o_wins:
        raise ValueError(
            "Impossible board: game should have ended after the game was won"
        )

    if x_wins and x_count != o_count + 1:
        raise ValueError(
            "Impossible board: game should have ended after the game was won"
        )

    if o_wins and x_count != o_count:
        raise ValueError(
            "Impossible board: game should have ended after the game was won"
        )

    if x_wins or o_wins:
        return "win"

    if x_count + o_count == 9:
        return "draw"

    return "ongoing"